modules/03_in_memory_storage/README.md

```diff
@@ -1,9 +1,471 @@
+            return None  # no words saved - nothing to return
+        return random.randint(0, len(self.storage) - 1)
+
+    def get_word_by_index(self, index: int) -> Optional[str]:
+        """
+        Given the index in the storage, return the secret word by this index.
+        """
+        if not (0 <= index < len(self.storage)):
+            return None  # index out of range - nothing to return
+        return self.storage[index]
+```
+
+And use the index instead of the word itself during the game:
+
+```python
+@app.route('/game', methods=['GET'])
+def game():
+    if 'secret_word_id' in session:
+        if database.get_word_by_index(session['secret_word_id']) is not None:
+            return render_template('game.html')
+
+    word_id = database.get_random_word_index()
+    if word_id is None:
+        flash("No words uploaded yet! Please upload at least one word to start guessing")
+        return redirect("/")
+
+    session['secret_word_id'] = word_id
+
+    return render_template('game.html')
+
+
+@app.route('/make_a_guess', methods=['POST'])
+def make_a_guess():
+    if 'secret_word_id' not in session:  # this should never happen
+        flash("You can't guess words without starting the game first!")
+        return redirect('/game')
+
+    secret_word = database.get_word_by_index(session['secret_word_id'])
+
+    if request.form['guessed_word'] == secret_word:
+        flash(Markup("You guessed right! Good job! The secret word was <b>%s</b>" % secret_word))
+        del session['secret_word_id']
+        return redirect('/')
+
+    flash("You didn't guess right! Try again!")
+    return redirect('/game')
+```
+
+Run the service, play the game, verify that it works.
+Decode the session cookie again, verify that now it contains just the index instead of the secret word itself. 
+
+### Part 5. Recap
+
+We've created a real mini-game. We learned how to:
+
+- upload data from an HTML page to the server: using forms and POST requests;
+- save data in memory to share it between all users of the server as long as the server is alive;
+- save data to user sessions to share it between requests of _the same user_;
+- read data from the session in the browser to demonstrate that you never should store secrets in the Flask session. 
+
+Now you can think how you can further improve this game! For example:
+
+- How to add hints to the game? I.e. a user would upload a word "red" with a hint "color of a tomato",
+  and on the game page you'd display the hint.
+- How to make the game more forgiving to spelling? I.e. when a secret word is "red", also accept a guess "Red".
+- How to let user know how close they are to the answer for them to not guess blindly? I.e. we can consider
+  a guess "rid" very close to "red" because they share many common letters. 
+- How to restrict cheating even more? Word indexes in the session are more secure than the words  
+  themselves, but eventually the player will learn what index corresponds to which word.
+
```

Comment:
```suggestion
- How to make the game more forgiving about spelling? I.e. when a secret word is "red", also accept the guess "Red".
- How would you ensure the same word can't be added to the list of words more than once?
- How to let the user know how close they are to the answer for them to not guess blindly? I.e. we can consider
  a guess "rid" very close to "red" because they share many common letters. 
- How to add hints to the game? I.e. a user would upload a word "red" with a hint "color of a tomato",
  and on the game page you'd display the hint.
- How would you display the list of the wrong guesses the user has made so far in the current game?
- How to restrict cheating even more? Word indexes in the session are more secure than the words  
  themselves, but eventually the player will learn what index corresponds to which word.
```